test/test_dt_utils.py

```diff
@@ -0,0 +1,129 @@
+THE SOFTWARE.
+"""
+
+import numpy as np
+
+from arraycontext import (  # noqa
+    pytest_generate_tests_for_pyopencl_array_context
+    as pytest_generate_tests
+)
+
+from grudge import DiscretizationCollection
+
+import grudge.op as op
+
+import pytest
+
+import logging
+
+
+logger = logging.getLogger(__name__)
+
+
+@pytest.mark.parametrize("name", ["interval", "box2d", "box3d"])
+def test_geometric_factors_regular_refinement(actx_factory, name):
+    from grudge.dt_utils import dt_geometric_factors
+
+    actx = actx_factory()
+
+    # {{{ cases
+
+    if name == "interval":
+        from mesh_data import BoxMeshBuilder
+        builder = BoxMeshBuilder(ambient_dim=1)
+    elif name == "box2d":
+        from mesh_data import BoxMeshBuilder
+        builder = BoxMeshBuilder(ambient_dim=2)
+    elif name == "box3d":
+        from mesh_data import BoxMeshBuilder
+        builder = BoxMeshBuilder(ambient_dim=3)
+    else:
+        raise ValueError("unknown geometry name: %s" % name)
+
+    # }}}
+
+    min_factors = []
+    for resolution in builder.resolutions:
+        mesh = builder.get_mesh(resolution, builder.mesh_order)
+        dcoll = DiscretizationCollection(actx, mesh, order=builder.order)
+        min_factors.append(op.nodal_min(dcoll, "vol", dt_geometric_factors(dcoll)))
+
+    # Resolution is doubled each refinement, so the ratio of consecutive
+    # geometric factors should satisfy: gfi+1 / gfi = 2
+    min_factors = np.asarray(min_factors)
+    ratios = min_factors[:-1] / min_factors[1:]
+    assert np.all(np.isclose(ratios, 2))
+
+
+@pytest.mark.parametrize("name", ["interval", "box2d", "box3d"])
+def test_non_geometric_factors(actx_factory, name):
+    from grudge.dt_utils import dt_non_geometric_factor, dt_geometric_factors
+
+    actx = actx_factory()
+
+    # {{{ cases
+
+    if name == "interval":
+        from mesh_data import BoxMeshBuilder
+        builder = BoxMeshBuilder(ambient_dim=1)
+    elif name == "box2d":
+        from mesh_data import BoxMeshBuilder
+        builder = BoxMeshBuilder(ambient_dim=2)
+    elif name == "box3d":
+        from mesh_data import BoxMeshBuilder
+        builder = BoxMeshBuilder(ambient_dim=3)
+    else:
+        raise ValueError("unknown geometry name: %s" % name)
+
+    # }}}
+
+    factors = []
+    geom_factors = []
+    degrees = list(range(1, 8))
+    for degree in degrees:
+        mesh = builder.get_mesh(1, degree)
+        dcoll = DiscretizationCollection(actx, mesh, order=degree)
+        factors.append(dt_non_geometric_factor(dcoll))
+        geom_factors.append(
+            op.nodal_min(dcoll, "vol", dt_geometric_factors(dcoll))
+        )
+
+    # Crude estimate, factors should behave like 1/N**2
+    factors = np.asarray(factors)
+    geom_factors = np.asarray(geom_factors)
+    lower_bounds = 1/(np.asarray(degrees)**2)
+    upper_bounds = geom_factors*lower_bounds
+
+    assert lower_bounds.all() <= factors.all() <= upper_bounds.all()
```

Comment:
I don't think this works. `.all()` returns an array of Booleans. Likewise, chained inequalities also only work on Python scalars, not arrays. I tried a quick fix, but the inequalities are violated.

Reply:
@inducer Ah whoops. I've updated it: https://github.com/inducer/grudge/pull/108/commits/77c2a0a256a69866b0b7233b91000c70942757b3